src/libopensc/pkcs15-pin.c

```diff
@@ -367,15 +367,19 @@ _sc_pkcs15_verify_pin(struct sc_pkcs15_card *p15card, struct sc_pkcs15_object *p
 		data.pin_reference = skey_info->key_reference;
 	}
 
-	if(p15card->card->reader->capabilities & SC_READER_CAP_PIN_PAD) {
-		if (!pincode && !pinlen)
-			data.flags |= SC_PIN_CMD_USE_PINPAD;
+	if(p15card->card->reader->capabilities & SC_READER_CAP_PIN_PAD && !pincode && !pinlen) {
+		data.flags |= SC_PIN_CMD_USE_PINPAD;
 
 		if (auth_info->attrs.pin.flags & SC_PKCS15_PIN_FLAG_SO_PIN)
 			data.pin1.prompt = "Please enter SO PIN";
 		else
 			data.pin1.prompt = "Please enter PIN";
 	}
+	else {
+		r = _validate_pin(p15card, auth_info, pinlen);
```

Comment:
Ok...have to take a look at it, because _validate_pin appears immediatly broken:

	/* if we use pinpad, no more checks are needed */
	if (p15card->card->reader->capabilities & SC_READER_CAP_PIN_PAD)
		return SC_SUCCESS;

This is testing if we have a pinpad, and not if we are actually using a pinpad (only if pinlen is 0). Have to check how this impacts the other callers.